pkg/tanka/find.go

```diff
@@ -2,123 +2,165 @@ package tanka
+			}
+		}()
 	}
 
-	// if path is a file, don't search deeper
-	if !stat.IsDir() {
-		return nil, nil
+	// push paths to channel
+	var pathMap = map[string]bool{} // prevent duplicates
+	for _, path := range paths {
+		if _, ok := pathMap[path]; ok {
+			continue
+		}
+		pathMap[path] = true
+		pathChan <- path
 	}
 
-	// list directory
-	files, err := os.ReadDir(path)
-	if err != nil {
-		return nil, findErr(path, err)
+	// collect jsonnet files
+	var jsonnetFiles []string
+	for i := 0; i < len(paths); i++ {
+		res := <-findJsonnetFilesChan
+		if res.err != nil {
+			return nil, res.err
+		}
+		jsonnetFiles = append(jsonnetFiles, res.jsonnetFiles...)
+	}
+	close(pathChan)
+	close(findJsonnetFilesChan)
+
+	findJsonnetFilesEndTime := time.Now()
+
+	// find all environments within jsonnet files
+	jsonnetFilesChan := make(chan string, len(jsonnetFiles))
+	findEnvsChan := make(chan findEnvsOut)
+
+	for i := 0; i < opts.Parallelism; i++ {
+		go func() {
+			for jsonnetFile := range jsonnetFilesChan {
+				// try if this has envs
+				list, err := List(jsonnetFile, Opts{JsonnetOpts: opts.JsonnetOpts})
+				if err != nil &&
+					// expected when looking for environments
+					!errors.As(err, &jpath.ErrorNoBase{}) &&
+					!errors.As(err, &jpath.ErrorFileNotFound{}) {
+					findEnvsChan <- findEnvsOut{err: fmt.Errorf("%s:\n %w", jsonnetFile, err)}
+					continue
+				}
+				filtered := []*v1alpha1.Environment{}
+				// optionally filter
+				if opts.Selector != nil && !opts.Selector.Empty() {
+					for _, e := range list {
+						if !opts.Selector.Matches(e.Metadata) {
+							continue
+						}
+						filtered = append(filtered, e)
+					}
+				} else {
+					filtered = append(filtered, list...)
+				}
+				findEnvsChan <- findEnvsOut{envs: filtered, err: nil}
+			}
+		}()
 	}
 
-	// it's not one. Maybe subdirectories are?
-	ch := make(chan findOut)
-	routines := 0
-
-	// recursively find in parallel
-	for _, fi := range files {
-		if !fi.IsDir() {
+	// push jsonnet files to channel
+	var jsonnetFileMap = map[string]bool{} // prevent duplicates
+	for _, jsonnetFile := range jsonnetFiles {
+		if _, ok := jsonnetFileMap[jsonnetFile]; ok {
 			continue
 		}
-
-		routines++
-		go findShim(filepath.Join(path, fi.Name()), opts, ch)
+		jsonnetFileMap[jsonnetFile] = true
+		jsonnetFilesChan <- jsonnetFile
 	}
 
-	// collect parallel results
-	var errs []error
+	// collect environments
 	var envs []*v1alpha1.Environment
-
-	for i := 0; i < routines; i++ {
-		out := <-ch
-		if out.errs != nil {
-			errs = append(errs, out.errs...)
+	var errs []error
+	for i := 0; i < len(jsonnetFiles); i++ {
+		res := <-findEnvsChan
+		if res.err != nil {
+			errs = append(errs, res.err)
 		}
-
-		envs = append(envs, out.envs...)
+		envs = append(envs, res.envs...)
 	}
+	close(jsonnetFilesChan)
+	close(findEnvsChan)
 
 	if len(errs) != 0 {
-		return envs, errs
+		return envs, ErrParallel{errors: errs}
```

Comment:
Similar comment about `errgroup`, although you just get the first error. Not sure how important it is to return all of them here?

Reply:
It's nice to get all the errors because instead of fixing them one by one, you get the whole list of errors

Reply:
ya, that makes sense